### ISSUE TITLE
Think about making duplicate BUILTIN bindings a warning

### ISSUE THREAD

**Issue by guillaumebrunerie** [agda/agda#3579]
It seems currently impossible to use the `BUILTIN EQUALITY` pragma and import `Agda.Builtin.String` at the same time, because `Agda.Builtin.String` imports `Agda.Builtin.Equality` which already has a `BUILTIN EQUALITY`. Note that `Agda.Builtin.Equality._≡_` is not even in scope in the code below so it’s annoying that it fails.

```agda
open import Agda.Builtin.String

data _==_ {A : Set} (a : A) : A → Set where
  refl : a == a

{-# BUILTIN EQUALITY _==_ #-}
```
```
Duplicate binding for built-in thing EQUALITY, previous binding to
Agda.Builtin.Equality._≡_
when checking the pragma BUILTIN EQUALITY _==_
```
The reason I want to do that is that I would like to use the `Prop`-valued equality as my builtin one, and I want to use reflection (which imports `Agda.Builtin.String`). This seems reasonable and it used to work a month ago, but not anymore (I guess it has to do with https://github.com/agda/agda/pull/3456 which added a dependency on `Agda.Builtin.Equality` to `Agda.Builtin.String`.)

I think it should be possible to either rebind builtins, or give the user a way to not import them (using `hiding …` does not work because the identifier is not even in scope).

**Comment by gallais** [comment]
I can put the primitives depending on `Agda.Builtin.Equality` in `Agda.Builtin.X.Properties`
instead if that would work for you!

**Comment by guillaumebrunerie** [comment]
It would work for me, but maybe it would make it harder for other users to discover those new primitives?

**Comment by fredrikNordvallForsberg** [comment]
Wouldn't the obvious thing be to make "Duplicate binding" a warning (that could then be turned off with `-W noDuplicateBuiltin`)?

**Comment by gallais** [comment]
@fredrikNordvallForsberg Things like `rewrite` and nat literals are assuming there's
only one BUILTIN for `EQUALITY` and `NATURAL` respectively, I believe. So this may cause
internal errors or confusing behaviours.

**Comment by fredrikNordvallForsberg** [comment]
Yeah, preparing a test case to see what will happen. I'm hoping for confusing behaviour only, which I think would be fair if you are willing to ignore the warning.

**Comment by fredrikNordvallForsberg** [comment]
Actually making it a warning seems to work mostly okay, except for some glitches with multiple warnings during imports of such warnings. But I don't think it is wise to include this change into 2.6 without further exposure, so I would suggest we go with @gallais' easy fix for now, and postpone the warning until 2.6.1. 

(There are similar problems with e.g. `Agda.Builtin.Nat` importing `Agda.Builtin.Bool` and a user wanting to import the builtin `Nat`, but make their own BUILTIN `Bool` for instance. Perhaps not a common use case, but...)

**Comment by gallais** [comment]
Keeping the issue open with an updated title to think about turning duplicate BUILTIN
bindings errors into warnings for a future release.

**Comment by rwe** [comment]
I was about to make a very closely-related issue so I guess I'll comment on this one. The limitation of a single binding for each built-in seems counterintuitive to me: it's a similar sort of surprise as if Agda were to allow only a unique binding for other values…(e.g. "Can't set `y = 0` because `0` is already bound to `x`!")

This came up for me because I was trying to re-use some utilities from https://github.com/UlfNorell/agda-prelude while also importing https://github.com/HoTT/HoTT-Agda; both of which (reasonably) bind some symbol to `EQUALITY`, resulting in the following error:
```
Duplicate binding for built-in thing EQUALITY, previous binding to
.Agda.Builtin.Equality._≡_
when scope checking the declaration
  import HoTT
```

At least on the implementation side, the relevant code looks to be around here: https://github.com/agda/agda/blob/99f0a7a21f3bc2584673deecb7e4c8a2484205f4/src/full/Agda/TypeChecking/Monad/Builtin.hs#L77-L83 and [here](https://github.com/agda/agda/blob/99f0a7a21f3bc2584673deecb7e4c8a2484205f4/src/full/Agda/TypeChecking/Monad/Builtin.hs#L532-L566).

**Comment by UlfNorell** [comment]
> Can't set `y = 0` because `0` is already bound to `x`!

It's more like *"Can't set `x = 1` because `x` is already set to `0`"*. The builtins are used by the type checker in lots of places and it's not clear how allowing multiple bindings would affect that (how would the type checker know which one to pick?, and what are the consequences of picking different ones at different times?). For instance, reflection needs to know what the builtin list type is when quoting terms. If you have multiple bindings for lists and pick the wrong one you get illformed terms.

**Comment by gallais** [comment]
@erydo [As far as I can tell](https://github.com/HoTT/HoTT-Agda/search?l=Agda&q=BUILTIN) HoTT-Agda is not using anything exotic. So they could
re-export the content of the appropriate `Agda.Builtin.X` (potentially after renaming
some definitions) and that should make it compatible with the prelude & the stdlib.

I would have sent a PR myself but I can't seem to build the project.

**Comment by guillaumebrunerie** [comment]
@erydo @gallais The HoTT-Agda library hasn’t been given much love recently, so it indeed doesn’t build with the latest version of Agda anymore. I believe that the last version of Agda it was used with did not bind BUILTIN EQUALITY in the builtin files, so that’s why HoTT-Agda does it, but it should indeed just use Agda.Builtin.Equality now.

**Comment by rwe** [comment]
@guillaumebrunerie @gallais 	
> the last version of Agda it was used with did not bind BUILTIN EQUALITY in the builtin files, so that’s why HoTT-Agda does it, but it should indeed just use Agda.Builtin.Equality now.

That makes a lot of sense for why it is that way, thank you for the background!

The HoTT build issue is tracked here: https://github.com/HoTT/HoTT-Agda/issues/51 though is unfortunately beyond my understanding to help much with it. But did link it to the root cause and some other work towards its fix.

**Comment by rwe** [comment]
@UlfNorell Your explanation is starting to make sense to me. I might have been confusing my expectations between "builtins" with "primitives".

When I saw "builtin" or "primitive", I read: _for this symbol, augment it with some opaque definition or additional constraints/debug info/optimized implementation/etc provided by the compiler implementation itself_. If I now understand correctly, "builtin" means something more like _when the compiler needs to construct a list type, it should refer to *this* type"_. And "primitive" is somewhat as I understood before.

And so `BUILTIN` declarations are actually configuration for the compiler itself. Those builtins seem similar to symbols like [`__builtin_abs` in a C compiler](https://github.com/llvm-mirror/clang/blob/131ed145cd89454d4be4ac606f7d8562edb6f8f3/include/clang/Basic/Builtins.def#L111). But in such cases, although the user can refer to it and use it, they don't get to pick its original name…_that is_ the name.

In my limited understanding, allowing users or libraries to customize those bindings seems to only to admit user error. As in this case, different libraries choosing different names; but perhaps also introducing inconsistencies between the type definition the compiler expects and the one provided. Should that even be allowed? Is there a use-case for allowing libraries other than `Agda.Builtin` to configure those things? Is there a reason a user would not want to import `Agda.Builtin.Foo`?

**Comment by rwe** [comment]
Ok, to answer my own question a little bit, one example of a different binding that I see is:

HoTT
```agda
postulate
  _↦_ : ∀ {i} {A : Type i} → A → A → Type i

{-# BUILTIN REWRITE _↦_ #-}
```

vs the std-lib https://github.com/agda/agda-stdlib/blob/4a3745d095f3774165c354d7ea8e23d12b109267/src/Debug/Trace.agda
```agda
postulate
  trace-eq : ∀ {a} {A : Set a} (a : A) str → trace str a ≡ a

{-# BUILTIN REWRITE _≡_ #-}
{-# REWRITE trace-eq #-}
```

That seems like a fairly specialized case of instrumentation for debugging, but looks valid. Given the concern that builtins must be globally self-consistent within the type-checker, yet users may occasionally want to provide alternative definitions, I wonder if those bindings could be abstracted to a single location as a sort of "top-level compilation configuration". i.e. Define a module structure for the builtins, and just make that QName a compiler flag. For example, if types like this were shipped with the compiler as a minimal required definition:

(Pardon my very beginner code below)
```agda
module Agda.Builtin.Config where

record NatConfig (Nat : Set) : Set where
  field
    _+_ : Nat → Nat → Nat
    -- …etc

record Target {Nat' Eq' : Set} : Set where
  field
    Equality : EqualityConfig Eq'
    Nat : NatConfig Nat'
  -- …etc
```

And std-lib provided default bindings:
```agda
module Agda.StdLib.Target where

Default : Agda.Builtin.Config.Target
Default = record
  { Equality = Agda.Prelude.Equality._≡_
  ; Nat = record
    { _+_ = Agda.Prelude.Nat._+_
    ; _-_ = Agda.Prelude.Nat._-_
  -- etc…
    }
  }

Debug : Agda.Builtin.Config.Target
Debug = Default { {- override some things -} }
```

The compiler could take a flag to look at a module name, defaulting to something like `Agda.StdLib.Target.Default`. So instead of relying on `{-# BUILTIN NATPLUS _+_ #-}` being declared once in one file, the compiler would refer to `Agda.StdLib.DefaultTarget.Nat._+_` or `MyOtherConfig.Target.Nat._+_`. And if a user or library wants to provide a different definition, they would just provide another `Target` record definition that could be selected.

This would reduce the special syntax surface area (no more "BUILTIN" comments); would make apparent the expected types of those built-ins; and would prevent libraries from conflicting with each other, while still affording the ability to provide customized configurations for those builtins.

Potential difficulties: I don't know to what degree the compiler is able to resolve modules/records for those things before the builtins are themselves bound. Also, maybe more importantly, it might make it more difficult to e.g. _exclude_ builtins like "String" (maybe you don't _want_ to do things with strings). I wonder how to tell which builtins are needed for which features? Because apparently not _all_ builtins need to be bound, but apparently some of them do in certain cases?

**Comment by gallais** [comment]
Note that some users want to both:

* declare their own BUILTINs for some types (these declarations are conflicting with the stdlib)
* use the parts of the stdlib which do not depend on these BUILTINs

Example: https://github.com/agda/agda-stdlib/issues/393


Note also that some BUILTINs do not have a fixed type. EQUALITY for instance supports (among other things) all of the following:

```agda
data _≡_ {A : Set} (x : A) : A → Set where
 refl : x ≡ x

data _≡_ {a} {A : Set a} (x : A) : A → Set where
 refl : x ≡ x

data _≡_ {a} {A : Set a} (x : A) : A → Set a where
 refl : x ≡ x

data _≡_ {a b} {A : Set a} (x : A) : A → Set (a ⊔ b) where
 refl : x ≡ x

data _≡_ {A : Set} (x : A) : A → Prop where
 refl : x ≡ x

data _≡_ {a} {A : Set a} (x : A) : A → Prop a where
 refl : x ≡ x

data _≡_ {a b} {A : Set a} (x : A) : A → Prop (a ⊔ b) where
  refl : x ≡ x
```

**Comment by gallais** [comment]
> I believe that the last version of Agda it was used with did not bind BUILTIN EQUALITY in the builtin files,

Makes sense. The incompatibility between agda-prelude and agda-stdlib is the reason
why we introduced `Agda.Builtin.X` files IIRC.

**Comment by rwe** [comment]
Thank you for the examples @gallais!

It seems that the compiler must expect _some_ consistent type for those builtins, doesn't it, otherwise it wouldn't be able to use them? (At least in that: "the type has a single explicit parameter and a single dependent parameter." Or however that's correctly phrased).

Doing some digging, some aspects of those builtins like equality/refl appear to be enforced directly by the compiler: https://github.com/agda/agda/blob/d058070b09b3e50572a2a7b651019821c3d07f0e/src/full/Agda/TypeChecking/Rules/Builtin.hs#L656-L706

Is there a way of expressing that kind of constraint in Agda itself? Is there a way of generally stating things like "the type of a `BUILTIN EQUALITY` type" in a way that any of those definitions could be provided in a record?

Similarly for `Nat` and others, [`src/full/Agda/TypeChecking/Rules/Builtin.hs`](https://github.com/agda/agda/blob/d058070b09b3e50572a2a7b651019821c3d07f0e/src/full/Agda/TypeChecking/Rules/Builtin.hs#L84) there are specific checks:
https://github.com/agda/agda/blob/d058070b09b3e50572a2a7b651019821c3d07f0e/src/full/Agda/TypeChecking/Rules/Builtin.hs#L256-L257

So even if someone wants to define their own `Nat` type, they can only do that as long as it conforms to that meta-type that the compiler requires (one zero constructor and one successor constructor, i.e. Peano construction). So it's difficult to tell which aspects of those builtin definitions are truly usefully customizable.

For things like `Nat`, it would seem reasonable to say: "This is the compiler-provided built-in `Nat` type, and it's Peano encoded because that's how it needs to be. You're free to use your own type by providing instances of `Equiv Agda.BuiltIn.Nat YourNat`". (Though as far as seeming reasonable, I understand by now that I do _not_ understand the many use cases people are using these things for.)
